parsers/v.c

```diff
@@ -0,0 +1,2892 @@
+		{
+			readToken (token);
+			if (expectToken (token, TOKEN_OPEN_CURLY))
+				parseBlock (token, scope, true);
+			else
+				unreadToken (token);
+		}
+		else
+			unreadToken (token);
+	}
+	else
+		unreadToken (token);
+
+	PARSER_EPILOGUE ();
+}
+
+// match: 'match' access* fqident cont '{' [
+//     [[vtype | expr] [',' [vtype | expr]]* | 'else'] block
+// ]* '}'
+static void parseMatch (tokenInfo *const token, int scope)
+{
+	Assert (isKeyword (token, KEYWORD_match));
+	PARSER_PROLOGUE ("match");
+
+	vString *access = NULL;
+	readToken (token);
+    skipAccessAndReadToken (token, &access);
+    if (!isClose (token))
+    {
+	    parseExprList (token, scope, NULL, false);
+	    readToken (token);
+    }
+    if (!expectToken (token, TOKEN_OPEN_CURLY))
+	    skipToTokenFull (TOKEN_OPEN_CURLY, NULL, token);
+    if (isToken (token, TOKEN_OPEN_CURLY))
+    {
+	    readToken (token);
+	    do // match clauses
+	    {
+		    while (true) // match clause comma-list
+		    {
+			    if (isToken (token, TOKEN_TYPE, TOKEN_IDENT))
+				    parseFullyQualified (token, true);
+
+			    if (isToken (token, TOKEN_OPEN_CURLY, TOKEN_CLOSE_CURLY))
+				    break;
+			    else if (isInitialType (token))
+				    parseVType (token, NULL, scope, false, false);
+			    else if (isKeyword (token, KEYWORD_else))
+			    {
+				    readToken (token);
+				    break;
+			    }
+			    else
+				    parseExpression (token, scope, NULL);
+			    readToken (token);
+			    if (!isToken (token, TOKEN_COMMA))
+				    break;
+			    readToken (token);
+		    }
+		    if (expectToken (token, TOKEN_OPEN_CURLY))
+		    {
+			    parseBlock (token, scope, false);
+			    readToken (token);
+		    }
+		    else
+			    break;
+	    }
+	    while (!isToken (token, TOKEN_EOF, TOKEN_CLOSE_CURLY));
+	    if (!expectToken (token, TOKEN_CLOSE_CURLY))
+		    skipToToken (TOKEN_CLOSE_CURLY, NULL);
+    }
+
+    PARSER_EPILOGUE ();
+}
+
+// select: 'select' '{' [list block]* '}'
+static void parseSelect (tokenInfo *const token, int scope)
+{
+	Assert (isKeyword (token, KEYWORD_select));
+	PARSER_PROLOGUE ("select");
+
+	readToken (token);
+    if (expectToken (token, TOKEN_OPEN_CURLY))
+    {
+	    readToken (token);
+	    do // select clauses
+	    {
+		    if (!isToken (token, TOKEN_CLOSE_CURLY))
+			    parseExprList (token, scope, NULL, false);
+		    readToken (token);
+		    if (expectToken (token, TOKEN_OPEN_CURLY))
+		    {
+			    parseBlock (token, scope, false);
+			    readToken (token);
+		    }
+	    }
+	    while (!isToken (token, TOKEN_EOF, TOKEN_CLOSE_CURLY));
+	    expectToken (token, TOKEN_CLOSE_CURLY);
+    }
+    else
+	    unreadToken (token);
+
+    PARSER_EPILOGUE ();
+}
+
+// for: list? [';' expr]* block
+static void parseFor (tokenInfo *const token, int scope)
+{
+	Assert (isKeyword (token, KEYWORD_for, KEYWORD_Sfor));
+	PARSER_PROLOGUE ("for");
+
+	vString *access = NULL;
+	readToken (token);
+	skipAccessAndReadToken (token, &access);
```

Comment:
The buffer assigned to &acccess can be never freed.

Reply:
Install valgrind, and run `make units LANGUAGES=V VG=1`
```
  #valgrind-error:                        3
	parser-v.r/v-import
	parser-v.r/v-match
	parser-v.r/v-statements

```

See Units/parser-v.r/v-*.d/VALGRIND.tmp for an example.

Reply:
Fixed. Thanks for running valgrind checks! 😊  I did try to run it on my Mac, but it didn't work...